provider/digital_ocean.go

```diff
@@ -0,0 +1,279 @@
+		for _, change := range changes {
+			logFields := log.Fields{
+				"record": change.ResourceRecordSet.Name,
+				"type":   change.ResourceRecordSet.Type,
+				"action": change.Action,
+				"zone":   zoneName,
+			}
+
+			log.WithFields(logFields).Info("Changing record.")
+
+			if p.DryRun {
+				continue
+			}
+			switch change.Action {
+			case DigitalOceanCreate:
+				_, _, err := p.Client.CreateRecord(context.TODO(), zoneName,
+					&godo.DomainRecordEditRequest{
+						Data: change.ResourceRecordSet.Data,
+						Name: change.ResourceRecordSet.Name,
+						Type: change.ResourceRecordSet.Type,
+					})
+				if err != nil {
+					return err
+				}
+			case DigitalOceanDelete:
+				recordID, err := p.getRecordID(zoneName, change.ResourceRecordSet)
+				if err != nil {
+					return err
+				}
+				_, err = p.Client.DeleteRecord(context.TODO(), zoneName, recordID)
+				if err != nil {
+					return err
+				}
+			case DigitalOceanUpdate:
+				recordID, err := p.getRecordID(zoneName, change.ResourceRecordSet)
+				if err != nil {
+					return err
+				}
+				_, _, err = p.Client.EditRecord(context.TODO(), zoneName, recordID,
+					&godo.DomainRecordEditRequest{
+						Data: change.ResourceRecordSet.Data,
+						Name: change.ResourceRecordSet.Name,
+						Type: change.ResourceRecordSet.Type,
+					})
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// ApplyChanges applies a given set of changes in a given zone.
+func (p *DigitalOceanProvider) ApplyChanges(changes *plan.Changes) error {
+	combinedChanges := make([]*DigitalOceanChange, 0, len(changes.Create)+len(changes.UpdateNew)+len(changes.Delete))
+
+	combinedChanges = append(combinedChanges, newDigitalOceanChanges(DigitalOceanCreate, changes.Create)...)
+	combinedChanges = append(combinedChanges, newDigitalOceanChanges(DigitalOceanUpdate, changes.UpdateNew)...)
+	combinedChanges = append(combinedChanges, newDigitalOceanChanges(DigitalOceanDelete, changes.Delete)...)
+
+	return p.submitChanges(combinedChanges)
+}
+
+// newDigitalOceanChanges returns a collection of Changes based on the given records and action.
+func newDigitalOceanChanges(action string, endpoints []*endpoint.Endpoint) []*DigitalOceanChange {
+	changes := make([]*DigitalOceanChange, 0, len(endpoints))
+
+	for _, endpoint := range endpoints {
+		changes = append(changes, newDigitalOceanChange(action, endpoint))
+	}
+
+	return changes
+}
+
+func newDigitalOceanChange(action string, endpoint *endpoint.Endpoint) *DigitalOceanChange {
+	change := &DigitalOceanChange{
+		Action: action,
+		ResourceRecordSet: godo.DomainRecord{
+			Name: endpoint.DNSName,
+			Type: suitableType(endpoint),
+			Data: endpoint.Target,
+		},
+	}
+	return change
+}
+
+// getRecordID returns the ID from a record.
+// the ID is mandatory to update and delete records
+func (p *DigitalOceanProvider) getRecordID(zone string, record godo.DomainRecord) (int, error) {
+	zoneRecords, _, err := p.Client.Records(context.TODO(), zone, nil)
```

Comment:
this does not need to be refecthed on each record id. I believe list of records  can be retrieved in submitChanges() method for each zone and passed to his method